### PR TITLE
AB#3369 - add update, delete subscription in backend

### PR DESCRIPTION
**Pull request by leiye29**
### Description

1. provide update use email path
2. provide subscription delete path
3. provide subscription update path 

### Screenshots (if applicable)

image

### Checklist

- [x] I have tested the changes locally
- [ ] I have updated the documentation if necessary
- [ ] I have added/updated tests that prove my fix is effective or that my feature works
- [x] I have checked that my code follows the project's coding style by running `npm run format:check`
- [x] I have checked that my code contains no linting errors by running `npm run lint`
- [x] I have checked that my code contains no type errors by running `npm run typecheck`
- [x] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [x] I have checked that all e2e tests pass by running `npm run test:e2e`

### CODE REVIEW

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java [hunk 1]
```diff
@@ -98,6 +102,44 @@ public SubscriptionModel getSubscriptionById(
...
+	public void updateSubscriptionById(
+			@NotBlank(message = "userId must not be null or blank")
+			@Parameter(description = "The id of the user.", example = "00000000-0000-0000-0000-000000000000")
+			@PathVariable String userId,
+
+			@NotBlank(message = "subscriptionId must not be null or blank")
+			@Parameter(description = "The id of the subscription.", example = "00000000-0000-0000-0000-000000000000")
+			@PathVariable String subscriptionId,
+			
+			@Validated @RequestBody PreferredLanguageModel preferredLanguageModel) {
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java:117]: When PUT'ing to a `subscription` resource, the convention is to have the model be some representation (either full, or partial) of a subscription.

`SubscriptionUpdateModel` for example

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java [hunk 2]
```diff
@@ -98,6 +102,44 @@ public SubscriptionModel getSubscriptionById(
...
+		
+		userService.updateSubscriptionForUser(userId, subscriptionId, preferredLanguageModel);
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java:119]: You should:
- check that the user exists and throw a 404 (ResourceNotFoundException) if not
- check that the subscription exists and throw a 404 if not

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java [hunk 3]
```diff
@@ -98,6 +102,44 @@ public SubscriptionModel getSubscriptionById(
...
+			@Parameter(description = "The id of the subscription.", example = "00000000-0000-0000-0000-000000000000")
+			@PathVariable String subscriptionId) {
+		final var user = userService.getUserById(userId)
+			.orElseThrow(() -> new ResourceNotFoundException("No user with id=[%s] was found".formatted(userId)));
+
+		final var subscription = user.getSubscriptions().stream()
+			.filter(byId(subscriptionId)).findFirst()
+			.orElseThrow(() -> new ResourceNotFoundException("No subscription with id=[%s] was found".formatted(subscriptionId)));
+
+		userService.deleteSubscriptionForUser(userId, subscription);
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java:140]:
```suggestion
		userService.deleteSubscriptionForUser(userId, subscriptionId);
```

should be sufficient here.. allow the caller to send the minimal amount of data in the simplest form that is required to complete the operation.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java [hunk 4]
```diff
@@ -50,4 +53,16 @@ public UserModel getUserById(
 		return userModelAssembler.toModel(user);
 	}
 
+	@PutMapping({ "/{id}" })
+	@Operation(summary = "Update a user by ID")
+	@ApiResponse(responseCode = "204", description = "The request has been successfully processed.")
+	public void updateUserById(
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:59]: I don't know if we want to have a PUT method for the aggregate root.. it gives the consumer the ability to accidentally delete the entire entity graph for the user.

I would prefer that we implement PATCH for the user, and then we have more control over what the API consumer can change.

I would remove this method.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/PreferredLanguageModel.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+import ca.gov.dtsstn.cdcp.api.web.validation.LanguageCode;
+import io.swagger.v3.oas.annotations.media.Schema;
+import jakarta.validation.constraints.NotNull;
+
+@Immutable
+@JsonDeserialize(as = ImmutablePreferredLanguageModel.class)
+public interface PreferredLanguageModel {
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/PreferredLanguageModel.java:13]: As mentioned in another comment.. this should probably be `SubscriptionUpdateModel`

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 6]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
 	}
 
+	public void updateUser(String userId, UserModel userModel) {
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(userModel, "userModel is required; it must not be null or blank");
+
+		log.debug("Fetching user [{}] from repository", userId);
+		final var user = userRepository.findById(userId).orElseThrow();
+
+		if (user.getEmail().equals(userModel.getEmail()) == false) {
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 7]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
+	public void updateSubscriptionForUser(String userId, String subscriptionId, PreferredLanguageModel preferredLanguageModel) {
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(subscriptionId, "subscriptionId is required; it must not be null or blank");
+		Assert.notNull(preferredLanguageModel.getMsLanguageCode(), "preferredLanguageModel.getMsLanguageCode is required; it must not be null or blank");
+
+		log.debug("Fetching user [{}] from repository", userId);
+		final var user = userRepository.findById(userId).orElseThrow();
+		final var preferredLanguage = languageRepository.findByMsLocaleCode(preferredLanguageModel.getMsLanguageCode()).get();
+
+		final var subscription = user.getSubscriptions().stream()
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:162]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 8]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
+	}	
+
+	public void deleteSubscriptionForUser(String userId, Subscription subscription) {
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(subscription, "subscription is required; it must not be null or blank");
+
+		final var user = userRepository.findById(userId).orElseThrow(() -> new ResourceNotFoundException("No user with id=[%s] was found".formatted(userId)));
+		user.getSubscriptions().remove(subscriptionMapper.toEntity(subscription));
+		userRepository.save(user);
+	}
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:175]: fixed.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java (hunk 1 above)
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java (hunk 4 above)
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:55]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java (hunk 3 above)
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/PreferredLanguageModel.java (hunk 5 above)
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/PreferredLanguageModel.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java (hunk 2 above)
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 9]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
 		return userRepository.findById(id).map(userMapper::fromEntity);
 	}
 
+	public void updateUser(String userId, UserModel userModel) {
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(userModel, "userModel is required; it must not be null or blank");
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 10]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
+
+		if (user.getEmail().equals(userModel.getEmail()) == false) {
+			user.setEmail(userModel.getEmail());
+			user.setEmailVerified(false);
+		}
+
+		userRepository.save(user);
+	}
+
+	public void updateSubscriptionForUser(String userId, String subscriptionId, PreferredLanguageModel preferredLanguageModel) {
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 11]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(subscriptionId, "subscriptionId is required; it must not be null or blank");
+		Assert.notNull(preferredLanguageModel.getMsLanguageCode(), "preferredLanguageModel.getMsLanguageCode is required; it must not be null or blank");
+
+		log.debug("Fetching user [{}] from repository", userId);
+		final var user = userRepository.findById(userId).orElseThrow();
+		final var preferredLanguage = languageRepository.findByMsLocaleCode(preferredLanguageModel.getMsLanguageCode()).get();
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 12]
```diff
@@ -137,6 +140,46 @@ public Optional<User> getUserById(String id) {
...
+		subscription.setLanguage(preferredLanguage);
+			
+		userRepository.save(user);
+	}	
+
+	public void deleteSubscriptionForUser(String userId, Subscription subscription) {
+		Assert.hasText(userId, "userId is required; it must not be null or blank");
+		Assert.notNull(subscription, "subscription is required; it must not be null or blank");
+
+		final var user = userRepository.findById(userId).orElseThrow(() -> new ResourceNotFoundException("No user with id=[%s] was found".formatted(userId)));
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java (hunk 4 above)
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/UsersController.java:55]: As implemented, this isn't a JSON PATCH operation. We can fix it in a future PR, but please add a `TODO` so we don't forget.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserModel.java [hunk 13]
```diff
@@ -21,6 +21,8 @@ public class UserModel extends BaseResourceModel<UserModel> {
 
 	private List<UserAttributeModel> userAttributes;
 
+	private List<SubscriptionModel> subscriptions;
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserModel.java:24]: This should be removed. Subscriptions should be sent back to the consumer via HATEOAS links, instead.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserUpdateModel.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+import org.immutables.value.Value.Immutable;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import jakarta.validation.constraints.NotNull;
+
+@Immutable
+@JsonDeserialize(as = ImmutableUserUpdateModel.class)
+public interface UserUpdateModel {
+
+	@NotNull
```
Comment (gregory-j-baker) [inline backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/model/UserUpdateModel.java:12]:
```suggestion
	@NotBlank
```

**Review by gregory-j-baker — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 15]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
 		return userRepository.findById(id).map(userMapper::toDomainObject);
 	}
 
+	public void updateUser(String userId, String email) {
+		Assert.notNull(userId, "userId is required; it must not be null");
+		Assert.notNull(email, "email is required; it must not be null or blank");
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:142]: I don't see the change. Did you forget to commit?

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 16]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
...
+		final var subscription = user.getSubscriptions().stream()
+			.filter(byId(subscriptionId)).findFirst().orElseThrow();
+		subscription.setLanguage(preferredLanguage);
+			
+		userRepository.save(user);
+	}	
+
+	public void deleteSubscriptionForUser(String userId, String subscriptionId) {
+		Assert.notNull(userId, "userId is required; it must not be null");
+		Assert.notNull(subscriptionId, "subscriptionId is required; it must not be null or blank");
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:173]: I don't see the change. Did you forget to commit?

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 17]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
...
+			user.setEmail(email);
+			user.setEmailVerified(false);
+		}
+
+		userRepository.save(user);
+	}
+
+	public void updateSubscriptionForUser(String userId, String subscriptionId, String languageId) {
+		Assert.notNull(userId, "userId is required; it must not be null");
+		Assert.notNull(subscriptionId, "subscriptionId is required; it must not be null or blank");
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:157]: I don't see the change. Did you forget to commit?

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java [hunk 18]
```diff
@@ -98,6 +107,54 @@ public SubscriptionModel getSubscriptionById(
...
+			.orElseThrow(() -> new ResourceNotFoundException("No subscription with id=[%s] was found".formatted(subscriptionId)));
+
+		final var language = languageService.readByMsLocaleCode(subscriptionUpdateModel.getMsLanguageCode())
+				.orElseThrow(() -> new ResourceNotFoundException("No language with msLanguageCode=[%s] was found".formatted(subscriptionUpdateModel.getMsLanguageCode())));
+		
+		userService.updateSubscriptionForUser(userId, subscription.getId(), language.getId());
+	}
+
+	@DeleteMapping({ "/{subscriptionId}" })
+	@Operation(summary = "delete a subscriptions by ID")
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/web/v1/controller/SubscriptionsController.java:139]: I don't see this change. Did you forget to commit?

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 19]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
...
+		if (user.getEmail() == null || user.getEmail().equals(email) == false) {
+			user.setEmail(email);
+			user.setEmailVerified(false);
+		}
+
+		userRepository.save(user);
+	}
+
+	public void updateSubscriptionForUser(String userId, String subscriptionId, String languageId) {
+		Assert.notNull(userId, "userId is required; it must not be null");
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: I don't see this change. Did you forget to commit?

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 20]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
...
+
+		final var subscription = user.getSubscriptions().stream()
+			.filter(byId(subscriptionId)).findFirst().orElseThrow();
+		subscription.setLanguage(preferredLanguage);
+			
+		userRepository.save(user);
+	}	
+
+	public void deleteSubscriptionForUser(String userId, String subscriptionId) {
+		Assert.notNull(userId, "userId is required; it must not be null");
```
[the remark this answers is not in the document]
Comment (gregory-j-baker) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: I don't see this change.

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java:154]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

File: backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java [hunk 21]
```diff
@@ -137,6 +137,46 @@ public Optional<User> getUserById(String id) {
...
+			user.setEmailVerified(false);
+		}
+
+		userRepository.save(user);
+	}
+
+	public void updateSubscriptionForUser(String userId, String subscriptionId, String languageId) {
+		Assert.notNull(userId, "userId is required; it must not be null");
+		Assert.notNull(subscriptionId, "subscriptionId is required; it must not be null or blank");
+		Assert.notNull(languageId, "languageId is required; it must not be null or blank");
```
[the remark this answers is not in the document]
Comment (leiye29) [reply backend/src/main/java/ca/gov/dtsstn/cdcp/api/service/UserService.java]: fixed

**Review by gregory-j-baker — APPROVED** [review]

**Pull request merged**